backend/substrapp/compute_tasks/outputs.py

```diff
@@ -2,74 +2,116 @@
-    """Registers the performance on the orchestrator
-
-    Args:
-        client: An open orchestrator client.
-        task_key: The key of the compute task that produced this performances.
-        algo_key: The key of the algo that produced this performance.
-        perf: The performance value.
-    """
-    performance_obj = {
-        "compute_task_key": task_key,
-        "metric_key": algo_key,
-        "performance_value": perf,
-        "compute_task_output_identifier": identifier,
-    }
-    client.register_performance(performance_obj)
+# Temporarily determine model category based on output identifier
+OUTPUT_HEAD_MODEL_IDENTIFIERS = ["local", "head"]
+
+
+class OutputSaver:
+    def __init__(self, ctx: context.Context):
+        self._ctx = ctx
+
+    def save_outputs(self) -> None:
+        """Saves the task outputs"""
+        performances = [o for o in self._ctx.outputs if o.kind == orchestrator.AssetKind.ASSET_PERFORMANCE]
+        models = [o for o in self._ctx.outputs if o.kind == orchestrator.AssetKind.ASSET_MODEL]
+
+        for perf in performances:
+            self._save_performance(perf)
+
+        if models:
+            self._save_models(models)
+
+        if self._ctx.has_chainkeys:
+            directories.commit_dir(
+                self._ctx.directories, directories.TaskDirName.Chainkeys, directories.CPDirName.Chainkeys
+            )
+
+    def _save_performance(self, output: context.OutputResource):
+        logger.info("saving performances")
+        with get_orchestrator_client(self._ctx.channel_name) as client:
+            perf_path = os.path.join(self._ctx.directories.task_dir, output.rel_path)
+            perf = _get_perf(perf_path)
+            performance_obj = {
+                "compute_task_key": self._ctx.task.key,
+                "metric_key": self._ctx.algo.key,
+                "performance_value": perf,
+                "compute_task_output_identifier": output.identifier,
+            }
+            client.register_performance(performance_obj)
+
+    def _save_models(self, models: list[context.OutputResource]):
+        logger.info("saving models")
+
+        local_models = []
+        for model in models:
+            local_model = self._save_model_to_local_storage(model)
+            local_models.append(local_model)
+
+        try:
+            with get_orchestrator_client(self._ctx.channel_name) as client:
+                localrep_data_models = client.register_models({"models": local_models})
+        except Exception as exc:
+            for model in local_models:
+                self._delete_model(model["key"])
+            raise exc
+
+        for localrep_data in localrep_data_models:
+            localrep_data["channel"] = self._ctx.channel_name
+            localrep_serializer = ModelRepSerializer(data=localrep_data)
+            try:
+                localrep_serializer.save_if_not_exists()
```

Comment:
I don't know if it should be done in this pull request but it should be better to use a REST to register the model in substra-backend API.
You will surely have to create the post api in the model view to do that as it does not exist. 
It will remove the dependency between the Django database and the compute engine.

For instance, it has been done to register task profiling : 
 - post request :  https://github.com/Substra/substra-backend/blob/main/backend/substrapp/tasks/tasks_compute_task.py#L194-L207
 - post api :  https://github.com/Substra/substra-backend/blob/main/backend/api/views/task_profiling.py#L45-L50

Reply:
Yes, this is definitely where we want to go, but not in this PR.